src/dragonball/src/vcpu/vcpu_manager.rs

```diff
@@ -0,0 +1,1005 @@
+    pub fn pause_all_vcpus(&mut self) -> Result<()> {
+        self.pause_vcpus(&self.present_vcpus())
+    }
+
+    /// resume all vcpus
+    pub fn resume_all_vcpus(&mut self) -> Result<()> {
+        self.resume_vcpus(&self.present_vcpus())
+    }
+
+    /// exit all vcpus, and never restart again
+    pub fn exit_all_vcpus(&mut self) -> Result<()> {
+        self.exit_vcpus(&self.present_vcpus())?;
+        // clear all vcpu infos
+        self.vcpu_infos.clear();
+        // release io manager's reference manually
+        self.io_manager.take();
+
+        Ok(())
+    }
+
+    /// revalidate IoManager cache of all vcpus
+    pub fn revalidate_all_vcpus_cache(&mut self) -> Result<()> {
+        self.revalidate_vcpus_cache(&self.present_vcpus())
+    }
+
+    /// return all present vcpus
+    pub fn present_vcpus(&self) -> Vec<u8> {
+        self.vcpu_infos
+            .iter()
+            .enumerate()
+            .filter(|(_i, info)| info.handle.is_some())
+            .map(|(i, _info)| i as u8)
+            .collect()
+    }
+
+    /// Get available vcpus to create with target vcpu_count
+    /// Argument:
+    /// * vcpu_count: target vcpu_count online in VcpuManager.
+    /// Return:
+    /// * return available vcpu ids to create vcpu .
+    fn calculate_available_vcpus(&self, vcpu_count: u8) -> Vec<u8> {
+        let present_vcpus_count = self.present_vcpus_count();
+        let mut available_vcpus = Vec::new();
+
+        if present_vcpus_count < vcpu_count {
+            let mut size = vcpu_count - present_vcpus_count;
+            for cpu_id in 0..self.vcpu_config.max_vcpu_count {
+                let info = &self.vcpu_infos[cpu_id as usize];
+                if info.handle.is_none() {
+                    available_vcpus.push(cpu_id);
+                    size -= 1;
+                    if size == 0 {
+                        break;
+                    }
+                }
+            }
+        }
+
+        available_vcpus
+    }
+
+    /// Present vcpus count
+    fn present_vcpus_count(&self) -> u8 {
+        self.vcpu_infos
+            .iter()
+            .fold(0, |sum, info| sum + info.handle.is_some() as u8)
+    }
+
+    /// Configure single vcpu
+    fn configure_single_vcpu(
+        &mut self,
+        entry_addr: Option<GuestAddress>,
+        vcpu: &mut Vcpu,
+    ) -> std::result::Result<(), VcpuError> {
+        vcpu.configure(
+            &self.vcpu_config,
+            &self.vm_fd,
+            &self.vm_as,
+            entry_addr,
+            None,
+        )
+    }
+
+    fn create_vcpu(
+        &mut self,
+        cpu_index: u8,
+        request_ts: TimestampUs,
+        entry_addr: Option<GuestAddress>,
+    ) -> Result<()> {
+        info!("creating vcpu {}", cpu_index);
+        // We will reuse the kvm's vcpufd after first creation, for we can't
+        // create vcpufd with same id in one kvm instance.
+        let kvm_vcpu = match &self.vcpu_infos[cpu_index as usize].vcpu_fd {
```

Comment:
vcpu_infos capacity is initialized with `max_cpu_count`
And cpu_index sent here could not be larger than max_cpu_count ( checked in `create_vcpus` function)
So here cpu_index could not be larger than vcpu_info().len() :)

Reply:
The worry is that `vcpu_infos` vec is being initialised in another method and this method is "blindly trusting" it's paramters. I'm sure the current code is correct today, but it might not be tomorrow if the constructor changed somehow, so it's a potential (future) maintenance issue.

All I'm suggesting is replacing all the explicit indexing with calls to `.get()` as that can handle the error without panicking.

Reply:
Got that. I have added `get()` check for all the accesses for  `vcpu_infos`:)